Add ability to display intermediate result while the test is running

andriisoldatenko opened the pull request:
Closes #1315


https://terminalizer.com/view/fc96dd924634

Example of usage.

```
$ go run main.go run scripts.js

          /\      |‾‾| /‾‾/   /‾‾/
     /\  /  \     |  |/  /   /  /
    /  \/    \    |     (   /   ‾‾\
   /          \   |  |\  \ |  (‾)  |
  / __________ \  |__| \__\ \_____/ .io

  execution: local
     script: scripts.js
     output: -

  scenarios: (100.00%) 1 scenario, 1 max VUs, 10m30s max duration (incl. graceful stop):
           * default: 1 iterations for each of 1 VUs (maxDuration: 10m0s, gracefulStop: 30s)

Press R to get summary or esc to run Ctrl-C/Ctrl-Z

     http_req_blocked...........: avg=171.47ms min=171.47ms med=171.47ms max=171.47ms p(90)=171.47ms p(95)=171.47ms
     http_req_connecting........: avg=113.95ms min=113.95ms med=113.95ms max=113.95ms p(90)=113.95ms p(95)=113.95ms
     http_req_duration..........: avg=114.69ms min=114.69ms med=114.69ms max=114.69ms p(90)=114.69ms p(95)=114.69ms
     http_req_receiving.........: avg=1.97ms   min=1.97ms   med=1.97ms   max=1.97ms   p(90)=1.97ms   p(95)=1.97ms
     http_req_sending...........: avg=103µs    min=103µs    med=103µs    max=103µs    p(90)=103µs    p(95)=103µs
     http_req_tls_handshaking...: avg=0s       min=0s       med=0s       max=0s       p(90)=0s       p(95)=0s
     http_req_waiting...........: avg=112.62ms min=112.62ms med=112.62ms max=112.62ms p(90)=112.62ms p(95)=112.62ms
     http_reqs..................: 1 0.730317/s
     vus........................: 1 min=1 max=1
     vus_max....................: 1 min=1 max=1


     http_req_blocked...........: avg=171.47ms min=171.47ms med=171.47ms max=171.47ms p(90)=171.47ms p(95)=171.47ms
     http_req_connecting........: avg=113.95ms min=113.95ms med=113.95ms max=113.95ms p(90)=113.95ms p(95)=113.95ms
     http_req_duration..........: avg=114.69ms min=114.69ms med=114.69ms max=114.69ms p(90)=114.69ms p(95)=114.69ms
     http_req_receiving.........: avg=1.97ms   min=1.97ms   med=1.97ms   max=1.97ms   p(90)=1.97ms   p(95)=1.97ms
     http_req_sending...........: avg=103µs    min=103µs    med=103µs    max=103µs    p(90)=103µs    p(95)=103µs
     http_req_tls_handshaking...: avg=0s       min=0s       med=0s       max=0s       p(90)=0s       p(95)=0s
     http_req_waiting...........: avg=112.62ms min=112.62ms med=112.62ms max=112.62ms p(90)=112.62ms p(95)=112.62ms
     http_reqs..................: 1 0.461023/s
     vus........................: 1 min=1 max=1
     vus_max....................: 1 min=1 max=1

^C
running (00m04.2s), 0/1 VUs, 0 complete and 1 interrupted iterations
default ✗ [--------------------------------------] 1 VUs  00m04.2s/10m0s  0/1 iters, 1 per VU
WARN[0004] No script iterations finished, consider making the test duration longer

     data_received..............: 11 kB 2.6 kB/s
     data_sent..................: 76 B  18 B/s
     http_req_blocked...........: avg=171.47ms min=171.47ms med=171.47ms max=171.47ms p(90)=171.47ms p(95)=171.47ms
     http_req_connecting........: avg=113.95ms min=113.95ms med=113.95ms max=113.95ms p(90)=113.95ms p(95)=113.95ms
     http_req_duration..........: avg=114.69ms min=114.69ms med=114.69ms max=114.69ms p(90)=114.69ms p(95)=114.69ms
     http_req_receiving.........: avg=1.97ms   min=1.97ms   med=1.97ms   max=1.97ms   p(90)=1.97ms   p(95)=1.97ms
     http_req_sending...........: avg=103µs    min=103µs    med=103µs    max=103µs    p(90)=103µs    p(95)=103µs
     http_req_tls_handshaking...: avg=0s       min=0s       med=0s       max=0s       p(90)=0s       p(95)=0s
     http_req_waiting...........: avg=112.62ms min=112.62ms med=112.62ms max=112.62ms p(90)=112.62ms p(95)=112.62ms
     http_reqs..................: 1     0.239332/s
     vus........................: 1     min=1 max=1
     vus_max....................: 1     min=1 max=1

```

na-- commented:
Thanks, we'll take a look soon! Before that, can you forego the update of the `golang.org/x/sys` dependency, since it's just adding noise to this PR? We can update it in a separate PR later, together with other dependencies.

andriisoldatenko commented:
> Thanks, we'll take a look soon! Before that, can you forego the update of the `golang.org/x/sys` dependency, since it's just adding noise to this PR? We can update it in a separate PR later, together with other dependencies.

done.

andriisoldatenko commented:
@na-- @MStoykov @imiric 

**few notes:**
- i switched to another approach based on https://pkg.go.dev/golang.org/x/term
- removed https://github.com/eiannone/keyboard

still there are few small bugs, i'm trying to fix right now. in OSX it's still break terminal size... so 
![img](https://www.evernote.com/l/AMOMRxDmve1GsaRY68VNBb8Yw5LChXM3Hc8B/image.png)

Please let me know wdyt regarding new approach?

andriisoldatenko commented:
on separate note:
assuming that we want unbuffered input (without having to hit enter)?

na-- commented:
> on separate note:
> assuming that we want unbuffered input (without having to hit enter)?

I haven't looked at the new code, but yes, we'd like this feature to not require hitting enter after whatever trigger we've decided is written (`R` in this case). It's looking like it might prove to be too difficult to implement reasonably, so we could re-evaluate, we'll try to take a look at the code and play with it ourselves tomorrow.

andriisoldatenko commented:
> > on separate note:
> > assuming that we want unbuffered input (without having to hit enter)?
> 
> I haven't looked at the new code, but yes, we'd like this feature to not require hitting enter after whatever trigger we've decided is written (`R` in this case). It's looking like it might prove to be too difficult to implement reasonably, so we could re-evaluate, we'll try to take a look at the code and play with it ourselves tomorrow.

Few links might help when you try on your side:
- https://www.reddit.com/r/golang/comments/4hktbe/read_user_input_until_he_press_ctrlc/d2qquov/?utm_source=reddit&utm_medium=web2x&context=3
- https://stackoverflow.com/questions/48831750/unbuffered-input-manager-in-go
- https://groups.google.com/g/golang-nuts/c/NvD-pOTASIk

imiric commented:
Wow, so this turned out to be a can of worms... :sweat_smile: Sorry Andrii!

There's a couple of issues with the current implementation:

- Since it's switching the terminal to raw mode on start any output with newlines will be messed up, as you've found out. For me on xterm the running progress bar looks like:

  ```shell
  running (01.7s), 1/5 VUs, 9 complete and 0 interrupted iterations
                                                                   default   [==>-----------------------------------] 1/5 VUs  01.7s/20.0s
  
  ```

- The output after pressing `R` also looks like the bottom half of your screenshot to me. Pressing Ctrl+C runs the `defer` in `readInput` which restores it to cooked mode and the final summary looks fine in that case, but if I let the test exit normally then the final summary is borked, which is likely a race condition and the restore should be done before the summary is printed normally.

  We could address this `R` issue by switching to cooked mode before printing the summary, and then back again to raw mode to interpret the keystroke, but I don't see how the progress bar can be fixed.

Given that k6 constantly writes to stdout, we can't really switch to raw mode for this feature.

So how about this: interpret `Enter` in the default cooked mode and avoid switching to raw mode entirely. From my tests this gets interpreted as LF / ASCII 10 and works fine, with no output issues, and we don't need any external library or even `bufio` for this. I'm not sure how cross-platform this is and whether it's the same LF value on e.g. Windows terminals, but we can test it.

The obvious drawback of this is that we can't listen for any other key presses, but I'd argue that we really shouldn't anyway. `Enter` is not normally useful anyway, and is a friendlier binding than `Shift+r` to me.

WDYT?

MStoykov commented:
Thanks for the PR @AndriiChuzhynov :tada: 

quick thoughts before actually looking at the code:
- This not hitting enter turns out to too big of a pain IMO and given the problems we usually have with windows, IMO we will need to use something like [termbox](https://github.com/nsf/termbox-go) or [tcell](https://github.com/gdamore/tcell) to get it to work. This is ... way too complicated, but also gives us more stuff like graphs, but this is definitely WAY outside the scope of this issue. So my vote is to need to hit enter at least for now and fix it later.
- if you run the code with `-race` you will see some races that need to be fixed.
- the races are printed ... weirdly even with 227eaa93
- running with `-v` is botching the output as well, probably for the same reason as above
- even without races and with 227eaa93 or f376629 my terminal sometimes(most actually) still gets botched even if I just leave it without doing anything and I am on Linux with alacritty and tmux.
- the progressbar is of center and is "moving" leaving 'd'-s from the 'default' 
![image](https://user-images.githubusercontent.com/312246/108042952-2c1db500-7049-11eb-8e20-bceb5e35e334.png)


I have the feeling all of those visual artifacts are due to one problem, and fixing them IMO is more important than whether we will be hitting enter or not. I at least don't get my terminal botched if I `^c` so :shrug: 

additionally (although this is likely unfixable for now) if the scripts are read from the stdin as in `cat script.js | k6 run -` the feature doesn't work, which is expected, just something to be kept in mind. I am not certain something like termbox will help that either way :shrug:

andriisoldatenko commented:
@imiric @MStoykov thanks for review and testing! So i'll try to address issues and return with new commit.

andriisoldatenko commented:
@na-- @imiric @MStoykov
if fixed 1 of 2 RC by adding locks around HandleSummary` type:

```
engine.MetricsLock.Lock()
	summaryResult, err := runner.HandleSummary(globalCtx, &lib.Summary{
		Metrics:         engine.Metrics,
		RootGroup:       engine.ExecutionScheduler.GetRunner().GetDefaultGroup(),
		TestRunDuration: executionState.GetCurrentTestRunDuration(),
	})
	engine.MetricsLock.Unlock()
```
but than... i get stucked on `[io.Copy](https://github.com/loadimpact/k6/blob/a0f18e713ddbb6a201d8628ea98ec07d00243a7c/cmd/run.go#L469)` and `handleSummaryResult`

any ideas where to check to fix this RC:

but concerned how to fix showProgress one:
```
Write at 0x00c0015ec000 by goroutine 45:
  github.com/loadimpact/k6/cmd.showProgress.func2()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/cmd/ui.go:296 +0x1ef
  github.com/loadimpact/k6/cmd.showProgress()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/cmd/ui.go:363 +0x3c5
  github.com/loadimpact/k6/cmd.getRunCmd.func1.1()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/cmd/run.go:191 +0x2cd

Previous read at 0x00c0015ec000 by goroutine 46:
  github.com/loadimpact/k6/cmd.showProgress.func1()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/cmd/ui.go:288 +0x90
  github.com/loadimpact/k6/cmd.(*consoleWriter).Write()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/cmd/ui.go:73 +0x24a
  github.com/sirupsen/logrus.(*Entry).write()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/vendor/github.com/sirupsen/logrus/entry.go:280 +0x2d4
  github.com/sirupsen/logrus.Entry.log()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/vendor/github.com/sirupsen/logrus/entry.go:251 +0x404
  github.com/sirupsen/logrus.(*Entry).Log()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/vendor/github.com/sirupsen/logrus/entry.go:287 +0x167
  github.com/sirupsen/logrus.(*Entry).Warn()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/vendor/github.com/sirupsen/logrus/entry.go:308 +0x344
  github.com/loadimpact/k6/cmd.getRunCmd.func1.2()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/cmd/run.go:228 +0x2cb

Goroutine 45 (running) created at:
  github.com/loadimpact/k6/cmd.getRunCmd.func1()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/cmd/run.go:186 +0x10aa
  github.com/dop251/goja.(*Runtime).functionproto_call()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_function.go:109 +0x1a5
  github.com/dop251/goja.(*Runtime).functionproto_call-fm()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_function.go:102 +0x8d
  github.com/dop251/goja.(*vm)._nativeCall()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1817 +0x7fe
  github.com/dop251/goja.call.exec()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1789 +0x201a
  github.com/dop251/goja.(*call).exec()
      <autogenerated>:1 +0x64
  github.com/dop251/goja.(*vm).run()
      /Users/andrii/workspace/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:307 +0x15e
 
```

looks like we have global variable to lock `outMutex` 🤔 

also i found i can reproduce this bug from time to time in master branch ^^ using 0dd04a37

na-- reviewed (changes requested):
Thanks for fixing the data race and reporting the one we have on `master`! I've tested your latest code and it works, though it's apparent we have to go back to the drawing board with the feature itself... :disappointed: The UX of the current approach is not great, but the complexity and fragility of the alternative is also bad, even if we manage to get it working nicely without a full-blown ncurses-like CLI UI... I've added a note in https://github.com/loadimpact/k6/issues/1315 to that effect and tag the issue as `evaluation needed`.

The locking you've employed to get rid of the data race also deserves some further consideration and improvements. I'll try to explain, even though there's no point in doing the work for now, given that it doesn't seem likely we'll merge the feature soon because of the other problems with the whole feature. 

You are currently locking the whole `Engine` from processing new metrics and thresholds for the whole duration of `handleSummary()`, however `handleSummary()` can potentially take quite a long time to execute if it makes HTTP requests or something like that. If it takes a long time, various VUs can continue generating new metric samples and sending them to [the buffered metric `Samples` channel in the `Engine`](https://github.com/loadimpact/k6/blob/6bebdd632eceadf6f8b50d4c47814b3a1688e976/core/engine.go#L93). I think it has a size of 1000 by default, and if that ever overflows, VUs will be blocked, which would affect the load test execution negatively.

A better locking approach instead would be to only lock the `Engine.MetricsLock` only while we export the data we pass to `handleSummary`, but not during the actual execution of the callback. So, basically, wrap this line: https://github.com/loadimpact/k6/blob/6bebdd632eceadf6f8b50d4c47814b3a1688e976/js/runner.go#L312

Or do it inside of the actual `summarizeMetricsToObject()` function: https://github.com/loadimpact/k6/blob/6bebdd632eceadf6f8b50d4c47814b3a1688e976/js/summary.go#L160

But as I said, no need to fix this or spend any more time on improvements until we're sure what we want to do here. And despite the fact we're not going to merge the PR for now, the whole effort has been very useful in illuminating all of the tricky issues here, so thank you very much for that and sorry again for the unexpected complexity!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/k6?pullRequest=1849) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/k6?pullRequest=1849) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/k6?pullRequest=1849) it.</sub>

na-- commented:
I'll close this so it doesn't clutter the list with open PRs. We are not going to merge it as it is, for the reasons explained in https://github.com/grafana/k6/pull/1849#pullrequestreview-592342355, and any lessons learned can still be gleaned from a closed PR, given that it's linked in https://github.com/grafana/k6/issues/1315 (which I'll keep open).
